### PR TITLE
Fix Node Version Management

### PR DESCRIPTION
**Pull request by ryuyu**
Update the caching script to have better fallbacks.
Update node management to use actual correct versions and resolve paths correctly.

@Chuxel @BretJohnson @lostintangent

### CODE REVIEW

File: Tasks/CordovaBuild/node-setup.js [hunk 1]
```diff
@@ -4,32 +4,52 @@
...
     taskLibrary = require('./lib/vso-task-lib-proxy.js'),
     teambuild = require('taco-team-build');
 
 var nodeSetupPromise;
+
+var cwd = taskLibrary.getInput('cwd', /* required */ false);
+var tacoFile = path.join((cwd ? cwd : '.'), 'taco.json');
+
 var version = taskLibrary.getInput('cordovaVersion', /* required */ false);
+if (!version) {
```
Comment (Chuxel) [inline Tasks/CordovaBuild/node-setup.js]: We may want to test for "" here as well.

Comment (MSLaguana) [inline Tasks/CordovaBuild/node-setup.js]: The empty string is falsy, so if version is "" then !version is true.

File: Tasks/CordovaBuild/node-setup.js [hunk 2]
```diff
@@ -4,32 +4,52 @@
...
+    try {
+        var stats = fs.statSync(tacoFile);
+        if (stats && stats.isDirectory()) {
+            version = require(tacoFile)['cordova-cli'];
+            console.log('Cordova version set to ' + version + ' based on the contents of taco.json');
+        }
+    } catch (e) { }
+}
+
```
Comment (Chuxel) [inline Tasks/CordovaBuild/node-setup.js]: There is one more fallback we need to consider unfortunately - that's if an environment variable is set.  The variable varies by task name. 

CORDOVA_DEFAULT_VERSION

or

IONIC_DEFAULT_VERSION

or 

PHONEGAP_DEFAULT_VERSION

So, order of precedence is:

1. Cordova version specified by task
2. Cordova version in taco.json
3. Env var
4. else edge

It's the same logic that's in taco-team-build - we should consider rationalizing it when we pick up the other refactor task to reduce risk of breaking if vso-task-lib drops support for 0.12.7

File: Tasks/CordovaCommand/lib/node-manager.js [hunk 3]
```diff
@@ -48,11 +48,23 @@ function setupMaxNode(maxVersion, targetVersion, /*optional*/ installNpmOnWindow
...
     }
-    var dlNodeCommand = new taskLibrary.ToolRunner(taskLibrary.which('curl', true));
+    
+    var curlPath = taskLibrary.which('curl', true);
+    if (!curlPath) {
+        taskLibrary.error("curl was not found in PATH. curl needs to be in the path on Windows. You can get curl by installing the Git Command Line Tools (www.git-scm.com/downloads)");
+        return null;
+    }
+    
+    var dlNodeCommand = new taskLibrary.ToolRunner(curlPath);
```
Comment (Chuxel) [inline Tasks/CordovaCommand/lib/node-manager.js]: Same comment as Cordova Build

There is one more fallback we need to consider unfortunately - that's if an environment variable is set. The variable varies by task name.

CORDOVA_DEFAULT_VERSION

So, order of precedence is:

1. Cordova version specified by task
2. Cordova version in taco.json
3. Env var
 4. else edge

It's the same logic that's in taco-team-build - we should consider rationalizing it when we pick up the other refactor task to reduce risk of breaking if vso-task-lib drops support for 0.12.7

File: Tasks/CordovaBuild/node-setup.js (hunk 2 above)
Comment (ryuyu) [inline Tasks/CordovaBuild/node-setup.js]: I thought the general format for the env var was PACKAGENAMEALLCAP_DEFAULT_VERSION, so we would only need to check for CORDOVA_DEFAULT_VERSION in this case? or does the version of the other packages imply a certain cordova version? if that is the case, there needs to be additional logic here to convert the CLI version into a corresponding cordova version. The ionic and phonegap version numbers do not seem to line up with the Cordova ones.

**Pull request merged**